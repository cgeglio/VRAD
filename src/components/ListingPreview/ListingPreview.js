import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './ListingPreview.scss';
import { Link } from 'react-router-dom';


class ListingPreview extends Component {
  constructor() {
    super();
    this.state = { favorite: '' };
  }

  determineIcon = () => {
    const { listing } = this.props;
    return listing.favorite ? 'filled' : 'unfilled';
  };

  determineFavorite = () => {
    const { listing, addFavorite } = this.props;
    listing.favorite = !listing.favorite;
    this.setState({ favorite: listing.favorite });
    addFavorite(listing);
  }

  render() {
    const { listing, setCurrentListing } = this.props;
    return (
      <article>
        <img src={process.env.PUBLIC_URL + `/images/${listing.listing_id}_a.jpg`} alt="Listing" className="preview-img" />
        <div className="listing-preview">
          <h2 className="listing-name">{listing.name}</h2>
          <div className="buttons">
            <Link to={`/area/${listing.area_id}/listings/${listing.listing_id}`}><button type="button" onClick={() => { setCurrentListing(listing) }} className="details-btn">View Details</button></Link>
            <button type="button" onClick={() => { this.determineFavorite() }} className="favorite-btn"><img src={process.env.PUBLIC_URL + `/${this.determineIcon()}-heart.png`} alt="Favorite icon" className="favorite-img" /></button>
          </div>
        </div>
      </article>
    );
  }
}

export default ListingPreview;

ListingPreview.propTypes = {
  listing: PropTypes.object.isRequired,
  addFavorite: PropTypes.func.isRequired,
  setCurrentListing: PropTypes.func.isRequired,
};
