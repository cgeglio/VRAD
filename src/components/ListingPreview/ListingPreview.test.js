import React from 'react';
import { shallow } from 'enzyme';
import ListingPreview from './ListingPreview';

describe('ListingPreview', () => {
  let wrapper;
  let mockListing;
  const mockAddFavorite = jest.fn();
  const mockSetCurrentListing = jest.fn();

  beforeEach(() => {
    mockListing = { name: 'test name', listing_id: 5, favorite: false };
    wrapper = shallow(<ListingPreview
      key={mockListing.listiing_id}
      listing={mockListing}
      addFavorite={mockAddFavorite}
      setCurrentListing={mockSetCurrentListing}
    />);
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should start off with a default state', () => {
    expect(wrapper.state()).toEqual({ favorite: '' });
  });

  it('should set state and call addFavorite when determineFavorite is called', () => {
    wrapper.instance().determineFavorite();
    expect(wrapper.state()).toEqual({ favorite: true });
    expect(mockAddFavorite).toHaveBeenCalledWith(mockListing);
  });

  it('should decide if the favorite icon is filled or unfilled when determineIcon is called', () => {
    expect(wrapper.instance().determineIcon()).toEqual('unfilled');
  });

  it('should call the setCurrentListing method when the view details button is clicked', () => {
    wrapper.find('.details-btn').simulate('click');
    expect(mockSetCurrentListing).toHaveBeenCalledWith(mockListing);
  });

  it('should call the determineFavorite method when the favorite button is clicked', () => {
    wrapper.instance().determineFavorite = jest.fn();
    wrapper.find('.favorite-btn').simulate('click');
    expect(wrapper.instance().determineFavorite).toHaveBeenCalled();
  });
});
