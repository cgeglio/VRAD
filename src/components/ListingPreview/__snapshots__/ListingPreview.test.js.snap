// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingPreview should match the snapshot with all data passed in correctly 1`] = `
<article>
  <img
    alt="Listing"
    className="preview-img"
    src="/images/5_a.jpg"
  />
  <div
    className="listing-preview"
  >
    <h2
      className="listing-name"
    >
      test name
    </h2>
    <div
      className="buttons"
    >
      <Link
        to="/area/undefined/listings/5"
      >
        <button
          className="details-btn"
          onClick={[Function]}
          type="button"
        >
          View Details
        </button>
      </Link>
      <button
        className="favorite-btn"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="Favorite icon"
          className="favorite-img"
          src="/unfilled-heart.png"
        />
      </button>
    </div>
  </div>
</article>
`;
